File: ui-frameworks/shadcn/pkg-py/src/shinyshadcn/_layout.py
from __future__ import annotations

from typing import Literal

import shinyreact


def card(
    *children: object,
    title: str | None = None,
    class_: str | None = None,
) -> shinyreact.Node:
    """A card container with an optional header title.

    Args:
        *children: Child nodes rendered inside the card body.
        title: Optional card header text.
        class_: Extra CSS classes merged onto the root element.
    """
    return shinyreact.Node(
        type="shadcn:Card",
        props={"title": title, "className": class_},
        children=list(children),
    )


def tab(value: str, label: str) -> dict[str, str]:
    """A single tab trigger spec for :func:`tabs`.

    Args:
        value: Identifier for this tab (matches the active-tab input value).
        label: Text shown on the tab trigger.
    """
    return {"value": value, "label": label}


def tabs(
    input_id: str,
    tabs: list[dict[str, str]],
    *panels: object,
    selected: str | None = None,
    class_: str | None = None,
) -> shinyreact.Node:
    """A tabbed panel. ``tabs`` defines the triggers; ``panels`` are the content.

    Panels are matched to tabs positionally — the Nth panel renders under the
    Nth tab. Server reads ``input.<input_id>()`` as the active tab's value.

    Args:
        input_id: Shiny input id for the active tab (two-way).
        tabs: Tab trigger specs, built with :func:`tab`.
        *panels: One content node per tab, in the same order as ``tabs``.
        selected: Initially active tab value (defaults to the first tab).
        class_: Extra CSS classes merged onto the root element.
    """
    return shinyreact.Node(
        type="shadcn:Tabs",
        props={
            "input_id": input_id,
            "tabs": tabs,
            "selected": selected,
            "className": class_,
        },
        children=list(panels),
    )


def collapsible(
    input_id: str,
    *children: object,
    trigger_label: str = "Toggle",
    open: bool = False,
    class_: str | None = None,
) -> shinyreact.Node:
    """A disclosure: a trigger reveals/hides its children. Server reads
    ``input.<input_id>()`` as a boolean (open).

    Args:
        input_id: Shiny input id.
        *children: Content shown when open.
        trigger_label: Label on the toggle button.
        open: Initial open state.
        class_: Extra CSS classes merged onto the root element.
    """
    return shinyreact.Node(
        type="shadcn:Collapsible",
        props={
            "input_id": input_id,
            "trigger_label": trigger_label,
            "open": open,
            "className": class_,
        },
        children=list(children),
    )


def accordion_item(value: str, title: str) -> dict[str, str]:
    """An accordion section header for :func:`accordion` (value + title)."""
    return {"value": value, "title": title}


def accordion(
    input_id: str,
    items: list[dict[str, str]],
    *panels: object,
    type: Literal["single", "multiple"] = "single",
    selected: object | None = None,
    class_: str | None = None,
) -> shinyreact.Node:
    """A vertical accordion. ``items`` are the section headers; ``panels`` are the
    content, matched positionally. Server reads ``input.<input_id>()`` as the open
    value(s) — string for "single", list for "multiple".

    Args:
        input_id: Shiny input id.
        items: Section specs, built with :func:`accordion_item`.
        *panels: One content node per item, in the same order.
        type: "single" (one open) or "multiple".
        selected: Initially open value(s).
        class_: Extra CSS classes merged onto the root element.
    """
    return shinyreact.Node(
        type="shadcn:Accordion",
        props={
            "input_id": input_id,
            "items": items,
            "type": type,
            "selected": selected,
            "className": class_,
        },
        children=list(panels),
    )


def resizable(
    *children: object,
    orientation: Literal["horizontal", "vertical"] = "horizontal",
    panels: list[dict[str, object]] | None = None,
    handle: bool = True,
    class_: str | None = None,
) -> shinyreact.Node:
    """A resizable panel group. Children are placed in panels separated by drag handles.

    Args:
        *children: Content nodes — each goes into one resizable panel.
        orientation: "horizontal" (side-by-side) or "vertical" (stacked).
        panels: Optional list of ``{"default_size": %, "min_size": %}`` per panel.
        handle: Show the grip icon on the resize handle.
        class_: Extra CSS classes merged onto the root element.
    """
    return shinyreact.Node(
        type="shadcn:Resizable",
        props={
            "orientation": orientation,
            "panels": panels or [],
            "handle": handle,
            "className": class_,
        },
        children=list(children),
    )


def scroll_area(
    *children: object,
    height: str = "300px",
    orientation: Literal["vertical", "horizontal", "both"] = "vertical",
    class_: str | None = None,
) -> shinyreact.Node:
    """A scrollable container. Children are the scroll content.

    Args:
        *children: Content nodes rendered inside the scrollable area.
        height: CSS height string (e.g. ``"300px"``).
        orientation: "vertical", "horizontal", or "both".
        class_: Extra CSS classes merged onto the root element.
    """
    return shinyreact.Node(
        type="shadcn:ScrollArea",
        props={"height": height, "orientation": orientation, "className": class_},
        children=list(children),
    )

File: ui-frameworks/shadcn/pkg-r/R/layout.R
# shadcn layout components. Generated split; edit here, then roxygenise.

#' A card container with an optional header title.
#' @param ... Child nodes rendered inside the card body.
#' @param title Optional card header text.
#' @param class Extra CSS classes merged onto the root element.
#' @return A `shinyreact` node.
#' @export
shadcn_card <- function(..., title = NULL, class = NULL) {
  node("shadcn:Card", ..., props = list(title = title, className = class))
}

#' A single tab trigger spec for shadcn_tabs().
#' @param value Identifier for this tab (matches the active-tab input value).
#' @param label Text shown on the tab trigger.
#' @param ... Must be empty; forces later arguments to be passed by name.
#' @return A list describing the entry, consumed by its parent component.
#' @export
shadcn_tab <- function(value, label, ...) {
  rlang::check_dots_empty()
  list(value = value, label = label)
}

#' A tabbed panel. `tabs` defines the triggers; `...` panels are the content.
#'
#' Panels are matched to tabs positionally — the Nth panel renders under the
#' Nth tab. Server reads input$<input_id> as the active tab's value.
#'
#' @param input_id Shiny input id for the active tab (two-way).
#' @param tabs List of tab trigger specs, built with shadcn_tab().
#' @param ... One content node per tab, in the same order as `tabs`.
#' @param selected Initially active tab value (defaults to the first tab).
#' @param class Extra CSS classes merged onto the root element.
#' @return A `shinyreact` node.
#' @export
shadcn_tabs <- function(input_id, tabs, ..., selected = NULL, class = NULL) {
  node("shadcn:Tabs", ..., props = list(
    input_id  = input_id,
    tabs      = tabs,
    selected  = selected,
    className = class
  ))
}

#' A disclosure: a trigger reveals/hides its children. Server reads
#' input$<input_id> as a boolean (open).
#' @param input_id Shiny input id.
#' @param ... Content shown when open.
#' @param trigger_label Label on the toggle button (default "Toggle").
#' @param open Initial open state (default FALSE).
#' @param class Extra CSS classes merged onto the root element.
#' @return A `shinyreact` node.
#' @export
shadcn_collapsible <- function(input_id, ..., trigger_label = "Toggle", open = FALSE,
                               class = NULL) {
  node("shadcn:Collapsible", ..., props = list(
    input_id = input_id, trigger_label = trigger_label, open = open, className = class
  ))
}

#' A resizable panel group. Children go in panels separated by drag handles.
#' @param ... Content nodes — each goes into one resizable panel.
#' @param orientation "horizontal" (side-by-side, default) or "vertical".
#' @param panels Optional list of list(default_size=%, min_size=%) per panel.
#' @param handle Show the grip icon on the resize handle (default TRUE).
#' @param class Extra CSS classes merged onto the root element.
#' @return A `shinyreact` node.
#' @export
shadcn_resizable <- function(..., orientation = "horizontal", panels = list(),
                             handle = TRUE, class = NULL) {
  node("shadcn:Resizable", ..., props = list(
    orientation = orientation,
    panels      = panels,
    handle      = handle,
    className   = class
  ))
}

#' A scrollable container. Children are the scroll content.
#' @param ... Content nodes rendered inside the scrollable area.
#' @param height CSS height string (default "300px").
#' @param orientation "vertical", "horizontal", or "both".
#' @param class Extra CSS classes merged onto the root element.
#' @return A `shinyreact` node.
#' @export
shadcn_scroll_area <- function(..., height = "300px", orientation = "vertical", class = NULL) {
  node("shadcn:ScrollArea", ..., props = list(
    height = height, orientation = orientation, className = class
  ))
}

#' An accordion section header for shadcn_accordion() (value + title).
#' @param value Identifier for this section.
#' @param title Header text.
#' @return A list describing the entry, consumed by its parent component.
#' @export
shadcn_accordion_item <- function(value, title) {
  list(value = value, title = title)
}

#' A vertical accordion. `items` are section headers; `...` panels are the
#' content, matched positionally. Server reads input$<input_id> as the open value(s).
#' @param input_id Shiny input id.
#' @param items List of section specs, built with shadcn_accordion_item().
#' @param ... One content node per item, in the same order.
#' @param type "single" (one open) or "multiple".
#' @param selected Initially open value(s).
#' @param class Extra CSS classes merged onto the root element.
#' @return A `shinyreact` node.
#' @export
shadcn_accordion <- function(input_id, items, ..., type = "single", selected = NULL,
                             class = NULL) {
  node("shadcn:Accordion", ..., props = list(
    input_id = input_id, items = items, type = type, selected = selected,
    className = class
  ))
}
